### PR TITLE
logical sorting for grid in eventwindow

### LINKED ISSUE #1370: ActivityId is does not sort propertly
Sorting the list of 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12 as

1
10
11
12
2
3
4
5
6
7
8
9

is not helpful.   The correct sorting is 

01
02
03
04
05
06
07
08
09
10
11
12

image

### PR DESCRIPTION
**Pull request by rifatx**
Resolves #1370.

### CODE REVIEW

File: src/PerfView/Utilities/LogicalGridDataComparer.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+            var xVal = _propertyInfo.GetValue(x).ToString();
+            var yVal = _propertyInfo.GetValue(y).ToString();
+
+            if (DateTime.TryParse(xVal, out var xDt) && DateTime.TryParse(yVal, out var yDt))
+            {
+                return xDt.CompareTo(yDt);
+            }
+
+            if (NativeMethods.StrCmpLogicalW(xVal, yVal) is var res
```
Comment (brianrob) [inline src/PerfView/Utilities/LogicalGridDataComparer.cs:56]: Can the call to `StrCmpLogicalW` be replaced with `string.Compare`?

**Review by brianrob — COMMENTED** [review]
@rifatx, apologies, this got lost in the shuffle.  Overall, this is seems great - one open question.

File: src/PerfView/Utilities/LogicalGridDataComparer.cs (hunk 1 above)
Comment (rifatx) [reply src/PerfView/Utilities/LogicalGridDataComparer.cs:56]: No prob @brianrob.

Not directly. For instance, `string.Compare("10", "2")` returns -1, where 10 is greater than 2. So direct string comparison is not the way to go.

It's possible to use a readily available package such as [this](https://www.nuget.org/packages/NaturalSort.Extension) or implement our own algorithm, but I believe this has two downsides:

- `StrCmpLogicalW ` is used for sorting files in Windows Explorer, so it provides better experience. Other implementations _may_ have a different "sorting quality".
- .NET implementations _may not_ be as performant (didn't run benchmarks on them though).

Comment (brianrob) [reply src/PerfView/Utilities/LogicalGridDataComparer.cs:56]: Thanks @rifatx.  Makes sense.  I did a bit of quick testing and ran into a failure when running https://github.com/brianrob/tests/tree/master/managed/activitypaths.  I collected using `PerfView collect /Providers=*MySource`.  When I filter down to the `MySource` events, just display the `ActivityID` field, and try to sort it by clicking on the column header, I get this:

```
System.InvalidOperationException: Failed to compare two elements in the array. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Utilities.LogicalGridDataComparer`1.Compare(Object x, Object y) in C:\src\perfview\src\PerfView\Utilities\LogicalGridDataComparer.cs:line 48
   at System.Array.SorterObjectArray.SwapIfGreaterWithItems(Int32 a, Int32 b)
   at System.Array.SorterObjectArray.PickPivotAndPartition(Int32 lo, Int32 hi)
   at System.Array.SorterObjectArray.IntroSort(Int32 lo, Int32 hi, Int32 depthLimit)
   at System.Array.SorterObjectArray.IntrospectiveSort(Int32 left, Int32 length)
   --- End of inner exception stack trace ---
   at System.Array.SorterObjectArray.IntrospectiveSort(Int32 left, Int32 length)
   at System.Array.Sort(Array keys, Array items, Int32 index, Int32 length, IComparer comparer)
   at System.Collections.ArrayList.Sort(Int32 index, Int32 count, IComparer comparer)
   at System.Collections.ArrayList.Sort(IComparer comparer)
   at MS.Internal.Data.SortFieldComparer.SortHelper(ArrayList al, IComparer comparer)
   at MS.Internal.Data.DataExtensionMethods.Sort(IList list, IComparer comparer)
   at System.Windows.Data.ListCollectionView.PrepareLocalArray()
   at System.Windows.Data.ListCollectionView.RefreshOverride()
   at System.Windows.Data.CollectionView.RefreshInternal()
   at System.Windows.Data.CollectionView.RefreshOrDefer()
   at System.Windows.Data.ListCollectionView.set_CustomSort(IComparer value)
   at PerfView.EventWindow.<.ctor>b__2_3(Object sender, DataGridSortingEventArgs e) in C:\src\perfview\src\PerfView\EventViewer\EventWindow.xaml.cs:line 162
   at System.Windows.Controls.DataGrid.OnSorting(DataGridSortingEventArgs eventArgs)
   at System.Windows.Controls.DataGrid.PerformSort(DataGridColumn sortColumn)
   at System.Windows.Controls.Primitives.DataGridColumnHeader.OnClick()
   at System.Windows.Controls.Primitives.ButtonBase.OnMouseLeftButtonUp(MouseButtonEventArgs e)
   at System.Windows.Controls.Primitives.DataGridColumnHeader.OnMouseLeftButtonUp(MouseButtonEventArgs e)
   at System.Windows.UIElement.OnMouseLeftButtonUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.ReRaiseEventAs(DependencyObject sender, RoutedEventArgs args, RoutedEvent newEvent)
   at System.Windows.UIElement.OnMouseUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseTrustedEvent(RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs args, Boolean trusted)
   at System.Windows.Input.InputManager.ProcessStagingArea()
   at System.Windows.Input.InputManager.ProcessInput(InputEventArgs input)
   at System.Windows.Input.InputProviderSite.ReportInput(InputReport inputReport)
   at System.Windows.Interop.HwndMouseInputProvider.ReportInput(IntPtr hwnd, InputMode mode, Int32 timestamp, RawMouseActions actions, Int32 x, Int32 y, Int32 wheel)
   at System.Windows.Interop.HwndMouseInputProvider.FilterMessage(IntPtr hwnd, WindowMessage msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at System.Windows.Interop.HwndSource.InputFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
```

Can you please take a look?

Comment (rifatx) [reply src/PerfView/Utilities/LogicalGridDataComparer.cs:71]: Ok @brianrob, obviously null values were not handled properly. 585444ac3c838863994064ce700052fc425e5e8e should do it.

**Review by brianrob — APPROVED** [review]
LGTM.  Thank you @rifatx!

**Pull request merged**